flow/stopwatch_dl.cc

```diff
@@ -0,0 +1,123 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/stopwatch_dl.h"
+#include "display_list/dl_blend_mode.h"
+#include "display_list/dl_canvas.h"
+#include "display_list/dl_paint.h"
+#include "include/core/SkPath.h"
+#include "include/core/SkRect.h"
+
+namespace flutter {
+
+static const size_t kMaxSamples = 120;
+static const size_t kMaxFrameMarkers = 8;
+
+void DlStopwatchVisualizer::Visualize(DlCanvas* canvas,
+                                      const SkRect& rect) const {
+  DlPaint paint;
+
+  // Establish the graph position.
+  auto const x = rect.x();
+  auto const y = rect.y();
+  auto const width = rect.width();
+  auto const height = rect.height();
+
+  // Scale the graph to show time frames up to those that are 3x the frame time.
+  auto const one_frame_ms = stopwatch_.GetFrameBudget().count();
+  auto const max_interval = one_frame_ms * 3.0;
+  auto const max_unit_interval = UnitFrameInterval(max_interval);
+  auto const sample_unit_width = (1.0 / kMaxSamples);
+
+  // Erase all pixels.
+  {
+    paint.setColor(0x99FFFFFF);
+    canvas->DrawRect(rect, paint);
+  }
+
+  // Prepare a path for the data; we start at the height of the last point so
+  // it looks like we wrap around.
+  {
+    SkPath path;
+    path.setIsVolatile(true);
+    path.moveTo(x, rect.bottom());
```

Comment:
Rather than using a path, which we'll end up tessellating per frame, I would create a vertices object and add the rectangles to this.

Here is an example of turing a bunch of rectangles into a single vertices call: https://github.com/gskinnerTeam/flutter-wonderous-app/blob/d369891d1eef36b2a7dbefc1566dd61c644890ca/lib/ui/screens/artifact/artifact_search/time_range_selector/time_range_painter.dart#L16-L46

The line can also be approximated with a skinny rectangle. 

The goal of this is to have the most minimally invasive rendering for the performance monitor.

Reply:
Thanks, this was really interesting!